### PR TITLE
Resolve target platform variables during p2 assembly (#3040)

### PR DESCRIPTION
**Pull request by kysmith-csg**
This updates the p2-repository-plugin:assemble-repository mojo to resolve any variables in the target definition locations just as the target-platform-configuration is able to do.

I modified the existing tests for target variables to include this because they are related and should both work together.

### CODE REVIEW

File: tycho-p2-repository-plugin/src/main/java/org/eclipse/tycho/plugins/p2/repository/AssembleRepositoryMojo.java [hunk 1]
```diff
@@ -349,8 +353,9 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                             .flatMap(tpFile -> tpFile.getLocations().stream())
                             .filter(InstallableUnitLocation.class::isInstance).map(InstallableUnitLocation.class::cast)
                             .flatMap(iu -> iu.getRepositories().stream())
-                            .filter(iuRepo -> autoReferencesFilter.test(iuRepo.getLocation()))
-                            .map(iuRepo -> new RepositoryReference(null, iuRepo.getLocation(), true))
+                            .filter(iuRepo -> autoReferencesFilter.test(varResolver.resolve(iuRepo.getLocation())))
+                            .map(iuRepo -> new RepositoryReference(null, varResolver.resolve(iuRepo.getLocation()),
+                                    true))
```
Comment (HannesWell) [inline tycho-p2-repository-plugin/src/main/java/org/eclipse/tycho/plugins/p2/repository/AssembleRepositoryMojo.java:358]: Lets save the second variable substitution.
```suggestion
                            .map(iuRepo -> varResolver.resolve(iuRepo.getLocation())).filter(autoReferencesFilter)
                            .map(location -> new RepositoryReference(null, location, true))
```

**Review by HannesWell — COMMENTED** [review]
In general this looks good, just one suggestion below.

Furthermore it would be good to verify in the Test that the URL is properly resolved.
You could do it like in `RepoRefLocationP2RepositoryIntegrationTest`.

Is the plan to also apply this to Tycho 5? Because if you had opened the PR against the master it could have probably be back-ported automatically.

**Comment by kysmith-csg** [comment]
> Is the plan to also apply this to Tycho 5?

I guess so yeah. I'm just more used to fixing things the opposite way (lower branches first then merge-forward).

**Comment by HannesWell** [comment]
> > Is the plan to also apply this to Tycho 5?
> 
> I guess so yeah. I'm just more used to fixing things the opposite way (lower branches first then merge-forward).

Understandable. But for back-ports from Tycho 5 to 4 an automated process exists where you just have to apply a lable and some GH actions do the rest.
AFAIK for the opposite way there is only the manual process.

Can you please also update the test case that runs the modified test-build to check for the resolved reference as indicated in my previous comment?
Then this change would be perfectly fine for me.

**Review by HannesWell — APPROVED** [review]
Pefect. Thanks.
Can you please squash all the there commits into one with a suitable Message/description for the overall change?

**Comment by laeubi** [comment]
@kysmith-csg also please rebase this on top of master, Tycho 4 is more ore less in maintenance mode so we usually do not implement new features to Tycho 4 but rather implement in Tycho 5 branch (where we can change everything) and then decide if it is save to be backported or requires some adjustments (e.g. keeping a previous default), that way it can't happen that something is lost for Tycho 5.

**Comment by kysmith-csg** [comment]
> we usually do not implement new features to Tycho 4 but rather implement in Tycho 5 branch

I disagree that this is a new feature - it is most definitely a bug. But I agree that it should be on latest first then backported. I'm not super familiar with the system you guys have and didn't know they can be backported automatically. Maybe helpful to add to the contributing guide?

I'm not familiar at all with rebasing branches. I'll try to use the UI because seemingly I can change the branch. If it gets too messed up I'll just open a new PR.

**Comment by kysmith-csg** [comment]
Yeah. This is why we can't trust UIs :) remaking the PR...

**Comment by laeubi** [comment]
> I disagree that this is a new feature - it is most definitely a bug.

Whatever we classify it, we need it in both branches as the bug is present in both of them, we have an automated way to backport (compatible) thing from master > Tycho 4 branch so usually it is easier to show that the bug still exits on master and then let it automatically backport :-)

**Comment by HannesWell** [comment]
> Maybe helpful to add to the contributing guide?

I think so. Do you want to provide a PR?
There is already a [backporting-section](https://github.com/eclipse-tycho/tycho/blob/master/CONTRIBUTING.md#backporting) but that is more general



> Yeah. This is why we can't trust UIs :) remaking the PR...

You probably would have to 'move' your commit to the master branch locally and force-push it. Moving can be done by switching to the master branch and cherry-picking the commit.
Any way, your approach worked as well 👍🏽

**Pull request closed without merging**